### ISSUE TITLE
type-mismatch false positive on unsizing coercion

### ISSUE THREAD

**Issue by jDomantas** [rust-lang/rust-analyzer#13864]
**rust-analyzer version**: rust-analyzer version: 0.3.1334-standalone (74ae2dd30 2022-12-25)

**rustc version**: rustc 1.66.0 (69f9c33d7 2022-12-12)

**relevant settings**: in vscode's settings.json: `"rust-analyzer.diagnostics.experimental.enable": true, "rust-analyzer.checkOnSave.enable": false` (to opt-in into type checking diagnostics)

This code:

```rust
struct Foo<T: ?Sized> {
    leading: u32,
    trailing: T,
}

fn unsize(x: Box<Foo<[u32; 4]>>) -> Box<Foo<[u32]>> {
    x
}
```

compiles fine with rustc, but vscode shows type error from rust-analyzer:

![image](https://user-images.githubusercontent.com/8329015/210071236-2aa1b5c6-d64f-48b2-8a50-5eb98787e745.png)

**Comment by lowr** [comment]
The cause of this bug boils down to [this FIXME](https://github.com/rust-lang/rust-analyzer/blob/3033c3ddbfcb0e42084ada8931e88d11eb98dee4/crates/hir-ty/src/chalk_db.rs#L617). #13112 implements it and would resolve this, but it's blocked on another issue.

**Comment by ChayimFriedman2** [comment]
Fixed now.
